ext/sockets/tests/socket_set_option_timeo_error.phpt

```diff
@@ -0,0 +1,60 @@
+--TEST--
+socket_set_option() with SO_RCVTIMEO/SO_SNDTIMEO/SO_LINGER
+--EXTENSIONS--
+sockets
+--FILE--
+<?php
+$socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
+if (!$socket) {
+        die('Unable to create AF_INET socket [socket]');
+}
+$options_1 = array("sec" => 1, "usec" => "aaaaa");
+$options_2 = array("sec" => new stdClass(), "usec" => "1");
+$options_3 = array("l_onoff" => "aaaa", "l_linger" => "1");
+$options_4 = array("l_onoff" => "1", "l_linger" => []);
+$options_5 = array("l_onoff" => PHP_INT_MAX, "l_linger" => "1");
+
+try {
+	socket_set_option( $socket, SOL_SOCKET, SO_RCVTIMEO, new stdClass);
+} catch (\TypeError $e) {
+	echo $e->getMessage() . PHP_EOL;
+}
+
+try {
+	socket_set_option( $socket, SOL_SOCKET, SO_RCVTIMEO, $options_1);
+} catch (\TypeError $e) {
+	echo $e->getMessage() . PHP_EOL;
+}
+
+try {
+	socket_set_option( $socket, SOL_SOCKET, SO_SNDTIMEO, $options_2);
+} catch (\TypeError $e) {
+	echo $e->getMessage() . PHP_EOL;
+}
+try {
+	socket_set_option( $socket, SOL_SOCKET, SO_LINGER, "not good");
+} catch (\TypeError $e) {
+	echo $e->getMessage() . PHP_EOL;
+}
+try {
+	socket_set_option( $socket, SOL_SOCKET, SO_LINGER, $options_3);
+} catch (\TypeError $e) {
+	echo $e->getMessage() . PHP_EOL;
+}
+try {
+	socket_set_option( $socket, SOL_SOCKET, SO_LINGER, $options_4);
+} catch (\TypeError $e) {
+	echo $e->getMessage() . PHP_EOL;
+}
+try {
+	socket_set_option( $socket, SOL_SOCKET, SO_LINGER, $options_5);
+} catch (\ValueError $e) {
+	echo $e->getMessage() . PHP_EOL;
+}
+?>
+--EXPECTF--
+socket_set_option(): Argument #4 ($value) must be of type array, stdClass given
+
+Warning: Object of class stdClass could not be converted to int in %s on line %d
```

Comment:
This is the only "detail" I'm not sure about regarding BC: is it okay to drop support for `stdClass`, `ArrayObject` and purpose built "data classes" (and maybe even objects in general)?

Maybe wait what @Girgias (or others) have to say about this.

Reply:
If you are talking about the warning, I don't see how this is a BC breaks.
All the cases you mention *already* trigger this warning.

If you are talking about not supporting casting random objects to arrays, that is a recurring "problem" with PHP objects being structs, reference pointers, and everything in between. I don't think it is a problem is properly documented in UPGRADING for master, as *most* functions do not allow objects in those cases. Nor does ZPP unless one uses the special `Z_PARAM_ARRAY_OR_OBJECT(_*)` specifier (or `A`/`H` for non-fast ZPP).

Moreover, this causes issues when the value in question possibly needs to be interpreted like a `string`.

However, as this is a bugfix, I would do `if (zval != IS_ARRAY) { /* slow path */ if (zval != IS_OBJECT) { /* throw */ } else { /* get HT from props */ } }`